Unexport unavailable metrics in Prometheus

agoode opened the pull request:
This introduces a new option to the prometheus integration to automatically unexport metrics for unavailable entities.

<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->


When an entity becomes unavailable, this component will continue to report the entity's last value, until Home Assistant itself restarts, or the entity returns. These stale metrics can be hard to notice, especially when the particular metric rarely changes (or changes slowly).

The `entity_available` metric is provided to let queries filter out unavailable metrics, but this is slow with current versions of prometheus (see https://github.com/prometheus/prometheus/pull/9577). And regardless of performance issues, including `entity_available` increases the complexity of promql expressions and is easy to forget.

Now this component will automatically withdraw metrics when the entity becomes unavailable, which matches the behavior on restart and makes it easier to see missing metrics without using an `unless`.


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/34632

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

knyar commented on tests/components/prometheus/test_init.py:
```diff
@@ -1284,6 +1283,125 @@ async def test_entity_becomes_unavailable_with_export(
+    await hass.async_block_till_done()
+    body = await generate_latest_metrics(client)
+
+    # Check if unavailable metrics are deleted.
```
```suggestion
    # Check that metrics no longer exported for an unavailable entity.
```

knyar reviewed (approved):
This looks great, thank you.

Any thoughts on whether this should eventually be the default? Might be something best to decide & announce now, and change the default value from `true` to `false` a few releases down the line.

agoode commented:
Yes, I think it would be good for this to become the new default. How would we announce this change?

knyar reviewed (approved):
> Yes, I think it would be good for this to become the new default. How would we announce this change?

I've looked at the [developer site](https://developers.home-assistant.io/) but have not been able to find a recommended process for changing the default values of configuration variables.

My suggestion would be to leave a comment mentioning future change of default in the code and in the docs (as part of https://github.com/home-assistant/home-assistant.io/pull/34632). We'll remove the comment in the same PR that will flip the default value, and will mark it as a breaking change.

agoode commented:
Thanks, I've updated both PRs.

knyar reviewed (approved):
Thank you! Hopefully one of the Home Assistant maintainers will be able to review & merge this soon.

rcloran reviewed (approved):
Nice!

I think there's a good argument to be made that an interim config option isn't necessary. Anyone making queries against entities which could become unavailable has to handle both cases right now, in case the entity is ever unavailable at start, so I think that not emitting metrics for unavailable entities is the correct behaviour -- put another way, I can't think of a case where you'd want to emit a NaN.

That said, giving an opportunity for people to test this in "production" for a while with an easy path to revert is a good thing. I'm just not sure it outweighs the drawbacks :)

agoode commented:
Thanks! Yes, let me take a look at removing the option entirely. That would definitely simplify things.

agoode commented:
I've updated both PRs.

knyar commented:
I don't have any objections to just doing this with no transition period, but maybe it reaches the level of a "Breaking change" now?

I can imagine use cases for which the current behavior is useful - for example, if you have a sensor or another device that is only intermittently available, having its last-known state reported to Prometheus might be more helpful than having metrics regularly disappear and reappear. In the new world, one would need to apply one of the `*_over_time` functions at query time to achieve something similar.

agoode commented:
I'm going to move this to draft because I want to think a little more about a couple things.

agoode commented:
The problem with the current PR is that going unavailable will unexport ALL the metrics, but I think we want to keep most around, especially `entity_available`. It shouldn't be too bad to fix.

I remembered I have a dashboard that shows which entities are unavailable, using `entity_available`. I forgot we need it still!

rcloran commented:
Sorry, I did not intend to create churn with my comment 🙈. I'm happy to see this proceed in either direction.

@knyar my point was that someone writing a "proper" query against an intermittent metric would have to handle missing metrics *anyways*, as those might occur during restart. I absolutely might be missing something here, though.

agoode commented:
No problem! I just realized that the original PR had some unintended effects, based on a few misunderstandings I had.

agoode commented:
Ok, I think it's good now. The `state_change`, `entity_available`, and `last_updated_time_seconds` metrics now stay around, which completely matches how it appears at startup and lets us continue to observe them.

knyar commented on homeassistant/components/prometheus/__init__.py:
```diff
@@ -248,6 +242,24 @@ def handle_state(self, state: State) -> None:
 
+        if state.state in IGNORED_STATES:
+            ignored_metric_names = set()
+            for metric in (state_change, entity_available, last_updated_time_seconds):
+                for sample in cast(list[prometheus_client.Metric], metric.collect())[
+                    0
+                ].samples:
+                    ignored_metric_names.add(sample.name)
```
I don't quite follow why we need to look at samples here, rather than simply pass `(state_change, entity_available, last_updated_time_seconds)` to `_remove_labelsets`?

agoode replied:
Thank you for pointing out my confusion. The other comment helped a lot too.

knyar commented on homeassistant/components/prometheus/__init__.py:
```diff
@@ -274,15 +286,25 @@ def handle_entity_registry_updated(
-                if sample.labels["entity"] == entity_id and (
-                    not friendly_name or sample.labels["friendly_name"] == friendly_name
+                if (
+                    sample.name not in ignored_metric_names
```
Could we skip based on the key in `self._metrics`? It's supposed to be the metric name anyway.

agoode replied:
Yes, this makes things much better. Thanks!

knyar commented on homeassistant/components/prometheus/__init__.py:
```diff
@@ -274,15 +286,25 @@ def handle_entity_registry_updated(
+        friendly_name: str | None = None,
+        ignored_metric_names: set[str] | None = None,
     ) -> None:
         """Remove labelsets matching the given entity id from all metrics."""
+        if ignored_metric_names is None:
```
The comment says "all metrics" which is no longer accurate, could you please adjust it?

joostlek reviewed (changes requested):
There's a merge conflict

agoode commented:
> There's a merge conflict

Thanks, I got a little stalled here but hope to get back to this soon.

Part of the issue is that the python client makes openmetrics-style metrics with additional suffixes for counters, so it's not super easy to work with them cleanly.

knyar commented on tests/components/prometheus/test_init.py:
```diff
@@ -1665,14 +1662,25 @@ async def test_disabling_entity(
 ) -> None:
-    """Test an entity that becomes unavailable is still exported."""
+    """Test an entity that becomes unavailable/unknown is no longer exported."""
+    reset_prometheus_registry()
+    config = {
+        prometheus.CONF_PROM_NAMESPACE: "",
+    }
+    assert await async_setup_component(
+        hass, prometheus.DOMAIN, {prometheus.DOMAIN: config}
+    )
+    await hass.async_block_till_done()
+    client = await hass_client()
```
I don't quite understand the reasons for resetting the registry and re-initializing the component here. I tried to [remove it](https://knyar.net/paste/384ded69), and the tests were still passing. Perhaps this was necessary when new behavior was behind a configuration flag, but it's not anymore?

agoode replied:
Yes, it was some kind of leftover code. I've removed it. Thanks!

knyar reviewed (approved):
Looks good, just one question about the tests.

agoode committed:
Prometheus: unexport unavailable metrics

When sensors go offline, this component would continue to report its
last value, until Home Assistant itself restarts, or the sensor
returns.

The `entity_available` metric can be used to filter out unavailable
metrics, but this is slow with current versions of prometheus
(see https://github.com/prometheus/prometheus/pull/9577).

Now, the component will automatically withdraw metrics when the entity
becomes unavailable, which matches the behavior on restart and
makes it easier to see missing metrics without using an `unless`.
